### ISSUE TITLE
ModuleConcatenation bailout: Module is not an ECMAScript module when inheriting exported based config

### ISSUE THREAD

**Issue by jdb8** [webpack/webpack#5408]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When inheriting from a base webpack config provided by an npm module and merging via webpack-merge, ModuleConcatenation fails to properly concatenate, giving the bailout warning 'Module is not an ECMAScript module' (despite the fact all files are indeed ECMAScript modules).

**If the current behavior is a bug, please provide the steps to reproduce.**
Created a simple repro repo here: https://github.com/jdb8/webpack-module-concat-repro. Instructions are detailed in README.md.

**What is the expected behavior?**
The output bundle should successfully concatenate the module with `ModuleConcatenationPlugin`. It shouldn't matter whether the plugin is defined in the base or child config, or whether the base config is exported from an npm module.

**If this is a feature request, what is motivation or use case for changing the behavior?**
N/A

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**
`webpack@3.4.1`, `node@5.11.0`

**Comment by jdb8** [comment]
Ok, so I jumped the gun in posting the description and title above - I should have tested my own repro steps on a fresh machine.

I've updated the repro to better describe the bug here: essentially, the module concatenation only fails if the version of webpack being `require`d in the base config module differs from the one in the child config (or perhaps the real issue is when it differs from the `webpack` binary being used to execute the config?).

Let me know if the new repro steps make sense and feel free to update the title to something more appropriate. I've definitely observed this kind of thing happening a lot (multiple versions of webpack being installed when using a non-flat installation method, and weird behaviour due to the `require`d versions differing from the binary). At a higher level, I think it'd be awesome if there was some way to guarantee that the `webpack` binary would always match the webpack lib files being imported into the config being executed, but I'm not sure how hard that'd be to enforce.

Even if this turns out not to be a huge bug in prod, it'd be cool to investigate this as a development QOL fix since this will likely crop-up whenever teams are working on inherited configs or during plugin development.

**Comment by matthewjh** [comment]
I had the same problem in #6402. Took a while to work it out.

I was running my build via a global install of webpack rather than the one in node_nodules.

**Comment by pigulla** [comment]
Just ran across this very issue. In my case the `es2015`-preset turned out to be the reason. See [this question](https://stackoverflow.com/questions/45384170/how-to-fix-modules-with-moduleconcatenation-bailout-module-is-not-an-ecmascrip) on StackOverflow.

**Comment by akaSybe** [comment]
any updates on this?

**Comment by trusktr** [comment]
Weird. My babel config is really simple, yet I still get this error.

Here's my Babel config:

```js
module.exports = {
    babelrc: false,
    plugins: [

        // Handle re-export syntaxes not yet supported by browsers or Webpack. f.e.
        // export * as foo from 'foo'
        // export default from 'foo'
        '@babel/plugin-proposal-export-namespace-from',
        '@babel/plugin-proposal-export-default-from',
        
        // Object rest spread will be supported in the upcoming Chromium-based
        // Edge, then we can drop it once it has enough adoption.
        '@babel/plugin-proposal-object-rest-spread',

    ],
}
```

And webpack modules config:

```js
    module: {
        rules: [
            {
                test: /\.jsx?$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: babelConfig,
                    },
                ],
            },
            {
                test: /\.tsx?$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: babelConfig,
                    },
                    {
                        loader: 'ts-loader',
                    },
                ],
            },
        ],
    },
```

And tsconfig `module` option is set to `es2015`.

So, I've no idea why on earth Webpack is telling me

```
ModuleConcatenation bailout: Module is not an ECMAScript module
```

Any idea how to see the code after the loaders are done handling it?

**Comment by trusktr** [comment]
Hmmm, seems like most of it was due to the entry point importing a CommonJS module, which caused the `not an ECMAScript module` error to subsequently occur for every other imported module. The error wasn't so helpful.

Then when I commented out that one line in my entry point which imported the CommonJS module, a bunch of stuff was able to concatenate.

Seems like error reporting could be more helpful in this area.
